### PR TITLE
WT-7757 Skip internal and leaf pages with content obsolete to the reader

### PR DESCRIPTION
**Pull request by sulabhM**
In WT-7619 we attempted to skip those leaf pages that have obsolete content as per the aggregated information for this particular reader. There is a cost to doing this check per leaf page when the same check done at a higher level could have helped in avoiding that entire sub-tree. In this change, I am using the function __wt_btcur_skip_page, as a custom skip function to the tree walk. This generalising the check over the whole walk, offers us an ability to skip larger portions of the tree. In my local reproduction of the reported performance issue, I see a significant performance improvement.

### CODE REVIEW

**Comment by sulabhM** [comment]
@keithbostic , @hbokhari ,

I am making this change so that we can walk past internal pages (sections of btree) if we know for this reader all the underlying content is obsolete. This follows on from WT-7619. Locally it offers some perf benefits for the delete and reinserts test that I am running.

Here is the issue, a stress run revealed a heap-after-use with __wt_page_is_modified. Below is the link to the failure. I know for WT-7619, it was discussed if we should check for modified after locking. Does this failure convince us to move the check inside lock? Could there be other faults in this change that trigger this heap-after-use ?

https://evergreen.mongodb.com/task_log_raw/wiredtiger_ubuntu1804_stress_tests_format_stress_sanitizer_smoke_test_1_patch_bb0ba80f60c45a10ff5bdb587efde5a602db99af_60fe14602a60ed49a6c7d66a_21_07_26_01_48_17/0?type=T

File: test/suite/test_cursor17.py
Comment (sulabhM) [inline test/suite/test_cursor17.py:100]: This test fails now because we are not visiting each leaf page and deciding to skip or not. We are skipping at a higher level, which means we are making less decisions on how many pages we skip, but lead to walking less leaf pages eventually. The test is written as a percentage of leaf pages seen vs skipped. @hbokhari , can you think of an alternative test scenario ?

File: dist/stat_data.py
Comment (sulabhM) [inline dist/stat_data.py:824]: I am not 100% sure how useful this statistic will be in its current form. Unless we know if we are skipping internal pages or leaf pages, we can't compare values and say skipping X is better than skipping Y. But it is still helpful in giving us an indication of when a search is skipping pages. I am not keen on adding 2 statistics here, one for leaf pages and the other for internal pages, considering there might be a performance impact with adding them here.

File: test/suite/test_cursor17.py
Comment (sulabhM) [reply test/suite/test_cursor17.py]: Discussed with @hbokhari and added a test to check that we skipped at least 1 page.

**Comment by sulabhM** [comment]
@keithbostic , this got finally passing. With the last commit, I am avoiding taking a local reference of a page, which would get split/evicted before the modified check got called. So now this is ready for review. I am not sure about the statistic  I added, would appreciate your opinion on the topic.

I tried another pull request with modified check outside the check, but page potentially gets evicted before the check gets called:
https://github.com/wiredtiger/wiredtiger/pull/6851

**Comment by keithbostic** [comment]
> @keithbostic , @hbokhari ,
>
> Here is the issue, a stress run revealed a heap-after-use with __wt_page_is_modified.

My recollection of the rule is you can't look at anything in the page without holding the `WT_REF` locked or a hazard pointer on the page. The walk function has done neither of those things when it calls the callback function, so only the information in the `WT_REF` structure is immediately available to that function.

File: src/btree/bt_curnext.c
Comment (keithbostic) [inline src/btree/bt_curnext.c:654]: There's a comment problem in this function, search for `not- found` (notice the space).

File: src/include/btree_inline.h
Comment (keithbostic) [inline src/include/btree_inline.h:2054]: This seems like a lot of work for something that can never be "true"? Why not pull the `session.txn.isolation` and `ignore_tombstone` checks to a higher level and don't specify a custom skip function if it's never going to be useful?

Comment (keithbostic) [reply src/include/btree_inline.h:2054]: If you pull the check for `WT_REF_MEM` to the set of `previous_state` and return immediately, you can outdent all of this one level and it will read more cleanly.

File: src/include/btree_inline.h
Comment (keithbostic) [inline src/include/btree_inline.h:2057]: Why only attempt to lock the page once, why not wait for the lock?

File: src/btree/bt_curnext.c
Comment (hbokhari) [inline src/btree/bt_curnext.c:681]: Removing this check means we can potentially read a page full of tombstones before the optimisation kicks in in the later part of this loop?

File: src/btree/bt_curnext.c
Comment (hbokhari) [inline src/btree/bt_curnext.c:756]: A comment around why we are using a custom page skip method here.

File: src/btree/bt_curprev.c
Comment (hbokhari) [inline src/btree/bt_curprev.c:704]: A comment around why we are using a page skip method here.

File: src/include/btree_inline.h
Comment (hbokhari) [reply src/include/btree_inline.h:2057]: For the benchmarks we have been looking at, we figured out that the cursor next and previous calls are the parts of the code where the application threads are spending more time. If we wait for the lock in a loop (i.e. use `WT_REF_LOCK()`), maybe it will introduce some regression. However, we are not sure as we don't have data about probability of failing to acquire the lock. 

@sulabhM - Maybe if we can add some debug stat to capture how many times we fail to acquire the lock, we will be able to decide between try-once or blocking wait. For example, not worth trying `WT_REF_LOCK()` if we are succeed the existing mechanism is able to lock the page 90% of the time.

File: src/btree/bt_curnext.c
Comment (sulabhM) [reply src/btree/bt_curnext.c:681]: yep, I think that is okay, being a single page that will avoid the optimization.

File: dist/stat_data.py
Comment (sulabhM) [reply dist/stat_data.py]: Agreed, removed the stat.

File: src/btree/bt_curnext.c
Comment (sulabhM) [reply src/btree/bt_curnext.c]: Fixed.

File: src/include/btree_inline.h
Comment (sulabhM) [reply src/include/btree_inline.h]: I have pulled out the checks specific to the walk outside of the skip function and deciding to call the tree walk with or without custom skip based on the checks. I have also pulled check for the previous state together with setting the locked state, and return early in case we do not get the lock.

Looks much cleaner now.

File: src/include/btree_inline.h
Comment (sulabhM) [reply src/include/btree_inline.h]: I will keep it as such for now. I will run a local test with block till locking, and see if it makes any difference. It will be hard to measure any perf difference we get comparing blocking vs try locking. I imagine a page is likely going through eviction/checkpoint we were unable to lock it, which means even if we decide to fall back to reading it, the read will have to wait. So, I suspect it wont make much difference in such a case.

File: src/btree/bt_curprev.c
Comment (sulabhM) [reply src/btree/bt_curprev.c]: Added a comment, resolving the conversation.

File: src/btree/bt_curnext.c
Comment (sulabhM) [reply src/btree/bt_curnext.c]: Added a comment, resolving the conversation.

File: src/btree/bt_curnext.c
Comment (hbokhari) [inline src/btree/bt_curnext.c:756]: The flag `WT_CURSTD_IGNORE_TOMBSTONE` is bit misleading. Here we are skipping the pages because we are not interested in returning a tombstones.

File: src/include/btree_inline.h
Comment (hbokhari) [inline src/include/btree_inline.h:2033]: Styling: comments should be on a separate line.

**Review by hbokhari — APPROVED** [review]
Apart from some minor fixes required in comments, the change looks good to me.

Make sure you have run the changes with MongoDB patch build.

File: src/include/btree_inline.h
Comment (sulabhM) [reply src/include/btree_inline.h]: Note: I did a few local runs on the workload I am investigating. I did not see any significant difference between blocking for a lock or giving up try locking.

File: src/btree/bt_curnext.c
Comment (sulabhM) [reply src/btree/bt_curnext.c]: Fixed.

File: src/include/btree_inline.h
Comment (sulabhM) [reply src/include/btree_inline.h:2033]: I picked this up from elsewhere in the code. We quite often using same line comments for variable initialization

File: src/btree/bt_curnext.c
Comment (keithbostic) [inline src/btree/bt_curnext.c:762]: I'm confused by this -- the comment says "and not interested in returning tombstones", but the code is `not WT_CURSTD_IGNORE_TOMBSTONE`.

I'd read the code as "if the WT_CURSTD_IGNORE_TOMBSTONE flag is set, we want to ignore tombstones",  which doesn't match the comment.

This appears in two  places, of course.

**Review by keithbostic — APPROVED** [review]

File: src/btree/bt_curnext.c
Comment (hbokhari) [reply src/btree/bt_curnext.c:762]: @keithbostic, the flag name, `WT_CURSTD_IGNORE_TOMBSTONE`, is the culprit. What this flags means is that, if set, caller is okay if the underlying cursor returns a value with globally visible tombstone. To the my best of knowledge, only history store cursor implementation applies this flag to the file cursor.

Maybe we can raise another ticket to think of a better name for this flag.

Comment (keithbostic) [reply src/btree/bt_curnext.c:762]: @hbokhari, I would definitely support a ticket to change the flag name -- "ignore tombstone means tombstones are OK" isn't intuitive to me! :)

Comment (sulabhM) [reply src/btree/bt_curnext.c:762]: Created WT-7915 to rename the flag to something more intuitive.

File: src/include/btree_inline.h
Comment (hbokhari) [inline src/include/btree_inline.h:2053]: @sulabhM, shouldn't this be `ref->page != NULL`?

Comment (sulabhM) [reply src/include/btree_inline.h:2053]: No, It is saying that if the page is NULL, don't have to proceed with the modified check, and instead can straight look at the aggregated address. So the page is modified check will only be performed if "ref->page == NULL" is false, ie "ref->page != NULL".

Comment (hbokhari) [reply src/include/btree_inline.h:2053]: Ah yes. Sorry I didn't realize that there is a "OR" condition there.

**Review by hbokhari — APPROVED** [review]
LGTM!

**Comment by sulabhM** [comment]
@keithbostic , please have another look. Stress tests pass now - with locking.

File: src/include/btree_inline.h
Comment (keithbostic) [reply src/include/btree_inline.h:2050]: @sulabhM, @hbokhari, I think there is a legitimate question being raised here. There are 5 possible `WT_REF.state` values: DISK, DELETED, LOCKED, MEM and SPLIT. So, what are the possible `previous_state` values?

* WT_REF_DISK    yes, do the test
* WT_REF_DELETED     yes, do the test
* WT_REF_LOCKED    can't happen, that would be a bad bug somewhere
* WT_REF_MEM    yes, do the test ONLY if the page hasn't been modified
* WT_REF_SPLIT    I am pretty confident this can't happen because the traversal code will have already dealt with split pages, but it's not as clear as other paths.

I think it might be a little safer (and future proof the code in case a new `WT_REF.state` was added in the future), and possibly add some clarity, to write the test this way:

```
if ((previous_state == WT_REF_DISK || previous_state == WT_REF_DELETED ||
    (previous_state == WT_REF_MEM && !__wt_page_is_modified(ref->page))) ...
```

I'm not testing `ref.page == NULL`, because if the page is in memory, the `ref.page` pointer has to be valid.

@sulabhM, this is an untested change, and I think the current code is actually OK, I just think this would add a bit of safety and maybe clarity to the code. Your call, and if you are good with the current code, that's fine with me.

**Review by keithbostic — APPROVED** [review]
@sulabhM, I pushed a couple of trivial changes (so hopefully this is ready to merge when you're next online). I don't think there's anything in there that requires more testing, and I don't think there's anything objectionable in there, but don't hesitate to revert without further discussion if you disagree.

If you want to sanity check d132e47, the traversal functions are indirecting through the `WT_REF` before calling `__wt_btcur_skip_page()`, it can't be NULL.

I'm raising one point in my the review for your consideration. I think that one condition could be rewritten for safety, or maybe clarity, but it's your call, the code as written is OK with me.

**Pull request merged**
